heclib/heclib_c/src/Public/ztsRetrieve.c

```diff
@@ -369,31 +370,208 @@ int ztsRetrieve(long long *ifltab, zStructTimeSeries *tss,
 		}
 		else {
-			status = ztsRetrieveIrreg7(ifltab, tss, 
+			status = ztsRetrieveIrreg7(ifltab, tss,
 									retrieveFlag, retrieveDoublesFlag, boolRetrieveQualityNotes);
-		}		
+		}
 		//  Do we need to get location information?
 		if ((status == STATUS_OKAY) && tss->locationStruct) {
 			zlocationRetrieve(ifltab, tss->locationStruct);
 		}
+		//--------------------------------------------------//
+		// convert to requested vertical datum if necessary //
+		//--------------------------------------------------//
+		char cPart[65];
+		zpathnameGetPart(tss->pathname, 3, cPart, sizeof(cPart));
+		if (!strncasecmp(cPart, "ELEV", 4)) {
+			//------------------------//
+			// parameter is elevation //
+			//------------------------//
+			char cvertical_datum[CVERTICAL_DATUM_SIZE];
+			int  ivertical_datum = -1;
+			verticalDatumInfo _vdi;
+			verticalDatumInfo *vdi = NULL;
+			char *vdiStr;
+			char errmsg[1024];
+			zquery("VDTM", cvertical_datum, sizeof(cvertical_datum), &ivertical_datum);
+			if (ivertical_datum != IVERTICAL_DATUM_UNSET) {
+				//-----------------------------------//
+				// specific vertical datum requested //
+				//-----------------------------------//
+				vdi = extractVerticalDatumInfoFromUserHeader(tss->userHeader, tss->userHeaderNumber);
+				if (!vdi) {
+					if (tss->locationStruct && tss->locationStruct->supplemental) {
+						vdiStr = extractFromDelimitedString(
+							&tss->locationStruct->supplemental,
+							VERTICAL_DATUM_INFO_USER_HEADER_PARAM,
+							":",
+							TRUE,
+							FALSE,
+							'\n');
+						if (vdiStr) {
+							char *msg = stringToVerticalDatumInfo(&_vdi, vdiStr);
+							if(msg == NULL) {
+								vdi = &_vdi;
+							}
+							free(vdiStr);
+						}
+					}
+				}
+				if (vdi == NULL) {
+					sprintf(
+						errmsg,
+						"Cannot convert from native vertical datum of '%s' to '%s'.\n"
+						"Record has no conversion information.\n"
+						"Values not converted.",
+						tss->locationStruct->verticalDatum == 0 ? "UNSET"   :
+						tss->locationStruct->verticalDatum == 1 ? "NAVD-88" :
+						tss->locationStruct->verticalDatum == 2 ? "NGVD-29" : "OTHER",
+						cvertical_datum);
+					status = zerrorProcessing(
+						ifltab,
+						DSS_FUNCTION_ztsRetrieve_ID,
+						zdssErrorCodes.INVALID_HEADER_PARAMETER,
+						intervalType,
+						0,
+						zdssErrorSeverity.WARNING,
+						tss->pathname,
+						errmsg);
+					return status;
+				}
+				else {
+					//---------------------------------------------------------------//
+					// ensure the vertical datum info is returned in the user header //
+					//---------------------------------------------------------------//
+					verticalDatumInfoToString(&vdiStr, vdi, TRUE);
+					char *headerString;
+					if (tss->userHeader) {
+						headerString = userHeaderToString(tss->userHeader, tss->userHeaderNumber);
+						if (!strstr(headerString, VERTICAL_DATUM_INFO_USER_HEADER_PARAM)) {
```

Comment:
Handle case where VDI is already in user header. Shouldn't happen, but if it does, the location VDI should be passed back